Tools/ftscalingbench/ftscalingbench.py

```diff
@@ -0,0 +1,322 @@
+        accu = next_number(i)
+    return accu
+
+def double(x):
+    return x + x
+
+module = sys.modules[__name__]
+
+@register_benchmark
+def module_function():
+    total = 0
+    for i in range(1000 * WORK_SCALE):
+        total += module.double(i)
+    return total
+
+class MyObject:
+    pass
+
+@register_benchmark
+def load_string_const():
+    accu = 0
+    for i in range(1000 * WORK_SCALE):
+        if i == 'a string':
+            accu += 7
+        else:
+            accu += 1
+    return accu
+
+@register_benchmark
+def load_tuple_const():
+    accu = 0
+    for i in range(1000 * WORK_SCALE):
+        if i == (1, 2):
+            accu += 7
+        else:
+            accu += 1
+    return accu
+
+@register_benchmark
+def create_pyobject():
+    for i in range(1000 * WORK_SCALE):
+        o = MyObject()
+
+@register_benchmark
+def create_closure():
+    for i in range(1000 * WORK_SCALE):
+        def foo(x):
+            return x
+        foo(i)
+
+@register_benchmark
+def create_dict():
+    for i in range(1000 * WORK_SCALE):
+        d = {
+            "key": "value",
+        }
+
+thread_local = threading.local()
+
+@register_benchmark
+def thread_local_read():
+    tmp = thread_local
+    tmp.x = 10
+    for i in range(500 * WORK_SCALE):
+        _ = tmp.x
+        _ = tmp.x
+        _ = tmp.x
+        _ = tmp.x
+        _ = tmp.x
+
+
+def bench_one_thread(func):
+    t0 = time.perf_counter_ns()
+    func()
+    t1 = time.perf_counter_ns()
+    return t1 - t0
+
+
+def bench_parallel(func):
+    t0 = time.perf_counter_ns()
+    for inq in in_queues:
+        inq.put(func)
+    for outq in out_queues:
+        outq.get()
+    t1 = time.perf_counter_ns()
+    return t1 - t0
+
+
+def benchmark(func):
+    delta_one_thread = bench_one_thread(func)
+    delta_many_threads = bench_parallel(func)
+
+    speedup = delta_one_thread * len(threads) / delta_many_threads
+    if speedup >= 1:
+        factor = speedup
+        direction = "faster"
+    else:
+        factor = 1 / speedup
+        direction = "slower"
+
+    use_color = hasattr(sys.stdout, 'isatty') and sys.stdout.isatty()
+    color = reset_color = ""
+    if use_color:
+        if speedup <= 1.1:
+            color = "\x1b[31m"  # red
+        elif speedup < len(threads)/2:
+            color = "\x1b[33m"  # yellow
+        reset_color = "\x1b[0m"
+
+    print(f"{color}{func.__name__:<18} {round(factor, 1):>4}x {direction}{reset_color}")
+
+def determine_num_threads_and_affinity():
+    if sys.platform != "linux":
+        return [None] * os.cpu_count()
+
+    # Try to use `lscpu -p` on Linux
+    import subprocess
+    try:
+        output = subprocess.check_output(["lscpu", "-p=cpu,node,core,MAXMHZ"], text=True)
+    except (FileNotFoundError, subprocess.CalledProcessError):
+        return [None] * os.cpu_count()
+
+    table = []
+    for line in output.splitlines():
+        if line.startswith("#"):
+            continue
+        cpu, node, core, maxhz = line.split(",")
```

Comment:
This line fails for me because `lscpu` returns 5 values for some reason:

```python
$ ./python Tools/ftscalingbench/ftscalingbench.py           
Traceback (most recent call last):
  File "/dev/cpython/Tools/ftscalingbench/ftscalingbench.py", line 322, in <module>
    main(options)
    ~~~~^^^^^^^^^
  File "/dev/cpython/Tools/ftscalingbench/ftscalingbench.py", line 289, in main
    initialize_threads(opts)
    ~~~~~~~~~~~~~~~~~~^^^^^^
  File "/dev/cpython/Tools/ftscalingbench/ftscalingbench.py", line 257, in initialize_threads
    cpus = determine_num_threads_and_affinity()
  File "/dev/cpython/Tools/ftscalingbench/ftscalingbench.py", line 227, in determine_num_threads_and_affinity
    cpu, node, core, maxhz = line.split(",")
    ^^^^^^^^^^^^^^^^^^^^^^
ValueError: too many values to unpack (expected 4, got 5)
```

This is the lscpu output:
```
$ lscpu --version
lscpu from util-linux 2.37.2
$ lscpu -p=cpu,node,core,MAXMHZ
# The following is the parsable format, which can be fed to other
# programs. Each different item in every column has an unique ID
# starting usually from zero.
# CPU,Node,Core,Maxmhz
0,0,0,3700,0000
1,0,0,3700,0000
2,0,1,3700,0000
...
```

Reply:
Is comma also used as the decimal separator on your system?

Reply:
Yes, it's because `lspcu` obeys LC_NUMERIC:

```
% LC_NUMERIC=nl_NL lscpu -p=cpu,node,core,MAXMHZ
# The following is the parsable format, which can be fed to other
# programs. Each different item in every column has an unique ID
# starting usually from zero.
# CPU,Node,Core,Maxmhz
0,0,0,3900,0000
1,0,1,3900,0000
2,0,2,3900,0000
3,0,3,3900,0000
```

(I would consider this a bug in `lscpu`, but there doesn't seem to be a way to disable this, so you'd have to make sure to clear all of `LANG`, `LC_ALL` and `LC_NUMERIC`.)

Reply:
Actually, explicitly passing `LC_NUMERIC=C` might be better.

Reply:
I confirm that this works now :slightly_smiling_face: